src/library/scala/collection/immutable/Range.scala

```diff
@@ -154,11 +154,12 @@ extends scala.collection.AbstractSeq[Int]
 
   @inline final override def foreach[@specialized(Unit) U](f: Int => U) {
     validateMaxLength()
-    val isCommonCase = (start != Int.MinValue || end != Int.MinValue)
+    val isCommonCase = start != Int.MinValue && start != Int.MaxValue && end != Int.MinValue && end != Int.MaxValue
     var i = start
     var count = 0
-    val terminal = terminalElement
     val step = this.step
+    val terminal = lastElement + step
+
```

Comment:
This second change is not necessary anymore: `terminalElement` *is* `lastElement + step`.